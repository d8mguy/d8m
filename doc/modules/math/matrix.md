## Matrix

Module for structural operations on matrices. Since 2d space types aren't implemented, these use 2d lists.

#### function transpose

    val transpose = \imp(mtx: list(list(T))) -> list(list(T)) {...}

