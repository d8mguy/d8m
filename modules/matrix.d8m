// Module for structural operations on matrices. Since 2d space types aren't implemented, these use 2d lists.

export transpose

given(T::entity) {
	val transpose = \imp(mtx: list(list(T))) -> list(list(T)) {
		if(mtx.count == 0) []
		else {
		    val nrows = mtx.count
		    val ncols = mtx[0].count
			var rslt = (0...ncols).{ val xx:list(T) = zerolist(nrows); xx }
			each(i^(0...nrows)) each(j^(0...ncols)) rslt[j][i] = mtx[i][j]
			rslt
		}
	}
}

// maybe add constructors for special matrix shapes and values; things like trace. But these quickly get numeric.
