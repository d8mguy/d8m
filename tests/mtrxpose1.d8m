// test transpose code in matrix module

import "matrix" as mtx
import go "math/rand"
val foo = \() {
    val randx = rand.New(rand.NewSource(1))
    val strgs = ["aa", "bb", "cc", "dd", "ee", "ff", "gg", "hh", "ii", "jj"]
    val row = \(n:integer) { (1..n).{ strgs[randx.Intn(40) % 10] } }
    val mtx1 = (1..7).{ row(10) }
    val tr1 = mtx.transpose(mtx1)
    println(mtx1)
    println(tr1)
}
foo()
