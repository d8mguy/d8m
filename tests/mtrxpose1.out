[[bb, hh, hh, jj, bb, ii, ff, aa, gg, aa], [ee, bb, cc, jj, ii, ee, bb, ff, hh, gg], [ff, gg, ii, ii, hh, hh, hh, ii, aa, ff], [bb, ii, hh, bb, jj, gg, hh, bb, ff, gg], [dd, aa, ee, dd, dd, hh, ii, ee, jj, dd], [hh, bb, jj, jj, aa, ff, ii, ii, dd, ff], [bb, aa, ff, gg, gg, ii, bb, cc, dd, gg]]
[[bb, ee, ff, bb, dd, hh, bb], [hh, bb, gg, ii, aa, bb, aa], [hh, cc, ii, hh, ee, jj, ff], [jj, jj, ii, bb, dd, jj, gg], [bb, ii, hh, jj, dd, aa, gg], [ii, ee, hh, gg, hh, ff, ii], [ff, bb, hh, hh, ii, ii, bb], [aa, ff, ii, bb, ee, ii, cc], [gg, hh, aa, ff, jj, dd, dd], [aa, gg, ff, gg, dd, ff, gg]]
